### PR TITLE
Add a Dockerfile for synapse

### PR DESCRIPTION
**Pull request by kaiyou**
*This is currently not ready for merge, the pull request is meant for tracking the progress on the branch.*

Changes
- [x] Dockerfile building properly
- [x] Basic Python start script
- [x] Template homeserver.yaml and log configuration
- [ ] Generate keys upon startup
- [ ] Handle shared secrets for TURN, registration, Macaroon
- [ ] Provide an example Docker Compose file with a separate TURN server

Issues
- [ ] SQL schema files are not copied properly when building

### CODE REVIEW

**Comment by kaiyou** [comment]
The image is working properly, I'll try and deploy it in production tonight. The PR still need a proper turn server working alongside with synapse in the example.

File: contrib/docker/README.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+to add parameters to the docker run command to account for the network situation
+with your postgres database.
+
+```
+docker run \
+    -d \
+    --name synapse \
+    -v ${DATA_PATH}:/data \
+    -e SYNAPSE_SERVER_NAME=my.matrix.host \
+    matrixdotorg/synapse:v0.22.1
```
Comment (turt2live) [inline contrib/docker/README.md]: This should probably be pointed at master, otherwise the version will be forgotten about and a bunch of old servers will be spun up.

File: contrib/docker/README.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+```
+
+
+## Volumes
+
+The image expects a single volume, located at ``/data``, that will hold:
+
+* temporary files during uploads;
+* uploaded media and thumbnais;
```
Comment (turt2live) [inline contrib/docker/README.md]: *thumbnails

File: contrib/docker/README.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+Unless you specify a custom path for the configuration file, a very generic
+file will be generated, based on the following environment settings.
+These are a good starting point for setting up your own deployment.
+
+Global settings:
+
+* ``UID``, the user id Synapse will run as [default 991]
+* ``GID``, the group id Synapse will run as [default 991]
+
+Synapse specific settings:
```
Comment (turt2live) [inline contrib/docker/README.md]: An environment variable for the cache factor would be neat, although most people aren't going to need it.

File: contrib/docker/README.md (hunk 1 above)
Comment (kaiyou) [reply contrib/docker/README.md]: Pointing at ``master`` sounds a bit risky as well. Maybe point to ``latest``, that the team will have a responsibility to point to the current working version?

Comment (turt2live) [reply contrib/docker/README.md]: master is effectively latest, but yes: latest would be safer.

Comment (kaiyou) [reply contrib/docker/README.md]: Oh yes, I had forgotten about that regarding this repo. But sure, let's go for ``latest`` anyway.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+RUN cd /synapse \
+ && pip install --upgrade pip setuptools psycopg2 \
+ && mkdir -p /synapse/cache \
+ && pip install -f /synapse/cache --upgrade --process-dependency-links . \
+ && mv /synapse/contrib/docker/* / \
+ && rm -rf setup.py setup.cfg synapse
+
+VOLUME ["/data"]
+
+EXPOSE 8448
```
Comment (uhoreg) [inline Dockerfile]: should also EXPOSE 8008

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+RUN apk add --no-cache --virtual .nacl_deps su-exec build-base libffi-dev zlib-dev openssl-dev libjpeg-turbo-dev linux-headers postgresql-dev
+
+COPY . /synapse
+
+# A wheel cache may be provided in ./cache for faster build
+RUN cd /synapse \
+ && pip install --upgrade pip setuptools psycopg2 \
+ && mkdir -p /synapse/cache \
+ && pip install -f /synapse/cache --upgrade --process-dependency-links . \
+ && mv /synapse/contrib/docker/* / \
```
Comment (uhoreg) [inline Dockerfile]: It seems strange to me to litter `/` with so many files.  Why not just keep them all there?

File: contrib/docker/conf/homeserver.yaml [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+#   require_transport_security: False
+#   notif_from: "Your Friendly %(app)s Home Server <noreply@example.com>"
+#   app_name: Matrix
+#   template_dir: res/templates
+#   notif_template_html: notif_mail.html
+#   notif_template_text: notif_mail.txt
+#   notif_for_new_users: True
+#   riot_base_url: "http://localhost/riot"
+
+enable_group_creation: true
```
Comment (uhoreg) [inline contrib/docker/conf/homeserver.yaml]: Would be nice to have some way to add appservices somehow (maybe in a future version of the Dockerfile).

File: contrib/docker/start.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+# Generate any missing shared secret
+for secret in ("SYNAPSE_REGISTRATION_SHARED_SECRET", "SYNAPSE_MACAROON_SECRET_KEY"):
+    if secret not in environ:
+        print("Generating a random secret for {}".format(secret))
+        environ[secret] = os.urandom(32).encode("hex")
+
+# Parse the configuration file
+if not os.path.exists("/compiled"): os.mkdir("/compiled")
+convert("/conf/homeserver.yaml", "/compiled/homeserver.yaml", environ)
+convert("/conf/log.config", "/compiled/%s.log.config" % environ.get("SYNAPSE_SERVER_NAME"), environ)
```
Comment (uhoreg) [inline contrib/docker/start.py]: I guess it should only do these if `SYNAPSE_CONFIG_PATH` is not set, and if `mode` is not `generate`

File: contrib/docker/conf/homeserver.yaml (hunk 6 above)
Comment (kaiyou) [reply contrib/docker/conf/homeserver.yaml]: Definitely agree, I was thinking maybe have all yaml files in a folder in ``/data`` be included as appservices.

File: Dockerfile (hunk 5 above)
Comment (kaiyou) [reply Dockerfile]: I usually prefer to keep short and simple path for files in Docker images, like ``/start.py`` and ``/conf``. But you are right, other files should not be copied.

File: contrib/docker/conf/homeserver.yaml (hunk 6 above)
Comment (kaiyou) [reply contrib/docker/conf/homeserver.yaml]: Just added this in the latest commit.

File: contrib/docker/README.md (hunk 3 above)
Comment (kaiyou) [reply contrib/docker/README.md]: The environment variable ``SYNAPSE_CACHE_FACTOR`` is supported by default if passed to the container. Just added some documentation about it, thanks :)

File: contrib/docker/start.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+import sys
+import subprocess
+import glob
+
+convert = lambda src, dst, environ: open(dst, "w").write(jinja2.Template(open(src).read()).render(**environ))
+mode = sys.argv[1] if len(sys.argv) > 1 else None
+environ = os.environ.copy()
+
+# Check mandatory parameters and build the base start arguments
+if "SYNAPSE_SERVER_NAME" not in environ:
```
Comment (jcgruenhage) [inline contrib/docker/start.py]: Doesn't this mean that one can never remove this env var? Even after the generation of the config?

File: contrib/docker/start.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        "--config-path", environ.get("SYNAPSE_CONFIG_PATH", "/compiled/homeserver.yaml")]
+
+# Generate any missing shared secret
+for secret in ("SYNAPSE_REGISTRATION_SHARED_SECRET", "SYNAPSE_MACAROON_SECRET_KEY"):
+    if secret not in environ:
+        print("Generating a random secret for {}".format(secret))
+        environ[secret] = os.urandom(32).encode("hex")
+
+# Load appservices configurations
+environ["SYNAPSE_APPSERVICES"] = glob.glob("/data/appservices/*.yaml")
```
Comment (jcgruenhage) [inline contrib/docker/start.py]: This would be additional work, but wouldn't it be a lot nicer to have folder based configuration for this? Have a path in the homeserver.yaml, in this case /data/appservices, and have synapse load all config files in there?

File: contrib/docker/conf/homeserver.yaml [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+# identifiers by this server.
+trusted_third_party_id_servers:
+    - matrix.org
+    - vector.im
+    - riot.im
+
+## Metrics ###
+
+enable_metrics: False
+report_stats: False
```
Comment (jcgruenhage) [inline contrib/docker/conf/homeserver.yaml]: Why is this hardcoded when there is an env var for this?

File: contrib/docker/README.md [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+  other options then).
+* ``SYNAPSE_NO_TLS``, set this variable to disable TLS in Synapse (use this if
+  you run your own TLS-capable reverse proxy).
+* ``SYNAPSE_WEB_CLIENT``, set this variable to enable the embedded Web client.
+* ``SYNAPSE_ENABLE_REGISTRATION``, set this variable to enable registration on
+  the Synapse instance.
+* ``SYNAPSE_ALLOW_GUEST``, set this variable to allow guest joining this server.
+* ``SYNAPSE_EVENT_CACHE_SIZE``, the event cache size [default `10K`].
+* ``SYNAPSE_CACHE_FACTOR``, the cache factor [default `0.5`].
+* ``SYNAPSE_REPORT_STATS``, set this variable to `yes` to enable anonymous
```
[the remark this answers is not in the document]
Comment (turt2live) [reply contrib/docker/README.md]: Defaulting to `yes` honestly feels like an argument waiting to happen. I'd somewhat recommend it be a required variable with no default.

File: contrib/docker/conf/homeserver.yaml (hunk 6 above)
Comment (kaiyou) [reply contrib/docker/conf/homeserver.yaml]: I want the image to have two different modes of operation: either generating and customizing a config file, or using the dnyamically generated one. I agree that the dynamically generated one could be populated using the generate command instead of Jinja inside the start script.

I'd say this is room for later improvement, as I'm not experimented enough to play with Synapse core code yet :)

File: Dockerfile [hunk 12]
```diff
@@ -0,0 +1,19 @@
+FROM python:2-alpine
```
[the remark this answers is not in the document]
Comment (kaiyou) [reply Dockerfile]: Disagreeing: I would care a lot more about having an updated Python than updated system libs as long as security patches are backported. docker.io/python:2-alpine is based on the latest Python release deployed separately from apk, as opposed to using the Alpine base image.

File: contrib/docker/docker-compose.yml [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    ports:
+      - 8448:8448/tcp
+    # ... or use a reverse proxy, here is an example for traefik:
+    labels:
+      - traefik.enable=true
+      - traefik.frontend.rule=Host:my.matrix.Host
+      - traefik.port=8448
+
+  db:
+    image: postgres:latest
```
[the remark this answers is not in the document]
Comment (kaiyou) [reply contrib/docker/docker-compose.yml]: Agreed.

File: contrib/docker/docker-compose.yml [hunk 14]
```diff
@@ -0,0 +1,38 @@
+# This compose file is compatible with Compose itself, it might need some
+# adjustments to run properly with stack.
+
+version: '3'
+
+services:
+
+  synapse:
+    image: matrixdotorg/synapse:latest
```
[the remark this answers is not in the document]
Comment (kaiyou) [reply contrib/docker/docker-compose.yml]: Agreed.

File: contrib/docker/docker-compose.yml [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+    # Since snyapse does not retry to connect to the database, restart upon
+    # failure
+    restart: unless-stopped
+    # See the readme for a full documentation of the environment settings
+    environment:
+      - SYNAPSE_SERVER_NAME=my.matrix.host
+      - SYNAPSE_ENABLE_REGISTRATION=yes
+    volumes:
+      - ./files:/data
```
[the remark this answers is not in the document]
Comment (kaiyou) [reply contrib/docker/docker-compose.yml]: Agreed.

File: contrib/docker/start.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+# In generate mode, generate a configuration, missing keys, then exit
+if mode == "generate":
```
[the remark this answers is not in the document]
Comment (kaiyou) [reply contrib/docker/start.py]: Disagreeing: again I see two modes of operation, using a dynamic config file based on environment variables or generate a fresh one from Synapse using ``generate`` and customize it manually. Maybe the documentation could be improved to explain this though.

File: contrib/docker/start.py [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+else:
+    # Parse the configuration file
+    if "SYNAPSE_CONFIG_PATH" not in environ:
+        if not os.path.exists("/compiled"): os.mkdir("/compiled")
+        convert("/conf/homeserver.yaml", "/compiled/homeserver.yaml", environ)
+        convert("/conf/log.config", "/compiled/%s.log.config" % environ.get("SYNAPSE_SERVER_NAME"), environ)
+    # Generate missing keys and start synapse
+    subprocess.check_output(args + ["--generate-keys"])
+    subprocess.check_output(["chown", "-R", permissions, "/data"])
+    os.execv("/sbin/su-exec", ["su-exec", permissions] + args)
```
[the remark this answers is not in the document]
Comment (kaiyou) [reply contrib/docker/start.py]: ``execv`` expects a path to the binary followed by the command line itself. ``su-exec`` is still the first argument of the command line (which will populate ``argv[0]``).

File: contrib/docker/start.py [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+convert = lambda src, dst, environ: open(dst, "w").write(jinja2.Template(open(src).read()).render(**environ))
+mode = sys.argv[1] if len(sys.argv) > 1 else None
+environ = os.environ.copy()
+
+# Check mandatory parameters and build the base start arguments
+if "SYNAPSE_SERVER_NAME" not in environ:
+    print("Environment variable SYNAPSE_SERVER_NAME is mandatory, exiting.")
+    sys.exit(2)
+
+permissions = "{}:{}".format(environ.get("UID", 991), environ.get("GID", 991))
```
[the remark this answers is not in the document]
Comment (kaiyou) [reply contrib/docker/start.py]: Agreed.

File: contrib/docker/conf/homeserver.yaml [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+
+## Database ##
+
+{% if POSTGRES_PASSWORD %}
+database:
+  name: "psycopg2"
+  args:
+    user: "{{ POSTGRES_USER or "matrix" }}"
+    password: "{{ POSTGRES_PASSWORD }}"
+    database: "{{ POSTGRES_DB or "matrix" }}"
```
[the remark this answers is not in the document]
Comment (kaiyou) [reply contrib/docker/conf/homeserver.yaml]: Agreed.

File: contrib/docker/conf/homeserver.yaml [hunk 20]
```diff
@@ -0,0 +1,213 @@
...
+listeners:
+  {% if not SYNAPSE_NO_TLS %}
+  -
+    port: 8448
+    bind_addresses: ['0.0.0.0']
+    type: http
+    tls: true
+    x_forwarded: false
+    resources:
+      - names: [client, webclient]
```
[the remark this answers is not in the document]
Comment (kaiyou) [reply contrib/docker/conf/homeserver.yaml]: Agreed.

File: contrib/docker/README.md [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+  key in order to enable recaptcha upon registration.
+* ``SYNAPSE_TURN_URIS``, set this variable to the coma-separated list of TURN
+  uris to enable TURN for this homeserver.
+* ``SYNAPSE_TURN_SECRET``, set this to the TURN shared secret if required.
+
+Shared secrets, these will be initialized to random values if not set:
+
+* ``SYNAPSE_REGISTRATION_SHARED_SECRET``, secret for registrering users if
+  registration is disable.
+* ``SYNAPSE_MACAROON_SECRET_KEY``, secret for Macaroon.
```
[the remark this answers is not in the document]
Comment (kaiyou) [reply contrib/docker/README.md]: I'd gladly add a Riot container to the Compose project later, which might (and probably will) have conflicting environment variables, won't it?

File: contrib/docker/README.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (kaiyou) [reply contrib/docker/README.md]: Agreed with @turt2live, will make it a required argument

File: contrib/docker/start.py (hunk 8 above)
Comment (kaiyou) [reply contrib/docker/start.py]: Yep, this got to be improved, will include the fix while making some other options mandatory. The expected behaviour would be: server name is mandatory for generating the config file, some variables are mandatory when using the dynamic config.

File: contrib/docker/conf/homeserver.yaml (hunk 10 above)
Comment (kaiyou) [reply contrib/docker/conf/homeserver.yaml]: My mistake, will fix.

File: Dockerfile (hunk 12 above)
[the remark this answers is not in the document]
Comment (jcgruenhage) [reply Dockerfile]: The people over at alpine are pretty fast at updating, and at the moment the versions of python in both images are the same. Looking over the python images again, a compromise might be `docker.io/python:2-alpine3.7`, which is basically the same image as `docker.io/python:2-alpine`, but is based on a 3.7. I am not sure why they don't default to that though, it doesn't really make sense to me.

File: contrib/docker/start.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (jcgruenhage) [reply contrib/docker/start.py]: I guess I should have read the documentation more carefully, I didn't get this thing with the dynamic config mode. That's a really cool feature!

File: contrib/docker/start.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (jcgruenhage) [reply contrib/docker/start.py]: Okay, good to know!

File: contrib/docker/README.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (jcgruenhage) [reply contrib/docker/README.md]: AFAIK this doesn't really matter, because in docker-compose, environment variables are defined per container, not per project.

File: contrib/docker/README.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (jcgruenhage) [reply contrib/docker/README.md]: This is fine too. I just feel like if it defaults to no, nearly no one will activate it.

File: contrib/docker/README.md [hunk 22]
```diff
@@ -76,12 +76,17 @@ Global settings:
...
 * ``UID``, the user id Synapse will run as [default 991]
 * ``GID``, the group id Synapse will run as [default 991]
+* ``SYNAPSE_CONFIG_PATH``, path to a custom config file
 
-Synapse specific settings:
+If ``SYNAPSE_CONFIG_PATH`` is set, you should generate a configuration file
+then customize it manually. No other environment variable is required.
+
+Otherwise, a dynamic configuration file will be used. The following environment
+variables are available for configuration:
```
Comment (jcgruenhage) [inline contrib/docker/README.md]: While it gets the point across fine, I'd put the two modes of operation somewhere more visible. People are likely to overlook it (buried in the env var list) and will likely be confused by this when things don't behave as they expect.

File: contrib/docker/README.md [hunk 23]
```diff
@@ -76,12 +76,17 @@ Global settings:
...
 
 * ``SYNAPSE_SERVER_NAME`` (mandatory), the current server public hostname.
-* ``SYNAPSE_CONFIG_PATH``, path to a custom config file (will ignore all
-  other options then).
+* ``SYNAPSE_REPORT_STATS``, (mandatory, ``yes`` or ``not``), enable anonymous
+  statistics reporting back to the Matrix project which helps us to get funding.
```
Comment (jcgruenhage) [inline contrib/docker/README.md]: Isn't it "no"?

File: contrib/docker/docker-compose.yml [hunk 24]
```diff
@@ -29,10 +34,13 @@ services:
       - traefik.port=8448
 
   db:
-    image: postgres:latest
+    image: postgres:10-alpine
```
Comment (jcgruenhage) [inline contrib/docker/docker-compose.yml]: I'd suggest the same as you already did above: prepend `docker.io/`
I'd also do it in the Dockerfile, for the alpine base, whichever that'll be (alpine or python-alpine).

File: contrib/docker/README.md [hunk 25]
```diff
@@ -12,7 +12,7 @@ use that server.
 Build the docker image with the `docker build` command from the root of the synapse repository.
 
 ```
-docker build -t matrixdotorg/synapse:v0.22.1 .
+docker build -t matrixdotorg/synapse .
```
Comment (jcgruenhage) [inline contrib/docker/README.md]: After the change in the compose file (the `docker.io/` prefix), you should probably change it here too. If someone builds the image and then tries to run the compose file, docker-compose will still try to pull in the image from docker.io

File: contrib/docker/docker-compose.yml [hunk 26]
```diff
@@ -6,16 +6,22 @@ version: '3'
...
   synapse:
-    image: matrixdotorg/synapse:latest
+    image: docker.io/matrixdotorg/synapse:latest
     # Since snyapse does not retry to connect to the database, restart upon
     # failure
     restart: unless-stopped
     # See the readme for a full documentation of the environment settings
     environment:
       - SYNAPSE_SERVER_NAME=my.matrix.host
+      - SYNAPSE_REPORT_STATS=no
```
Comment (jcgruenhage) [inline contrib/docker/docker-compose.yml]: The default "no" has just been moved now, I am still not sure people will take a serious look at this.

**Comment by jcgruenhage** [comment]
Nice, I put up another review and it's instantly outdated because I started before the new commit 🤣

File: contrib/docker/README.md (hunk 22 above)
Comment (kaiyou) [reply contrib/docker/README.md]: Just pushed some documentation update about this.

File: contrib/docker/README.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (kaiyou) [reply contrib/docker/README.md]: Okay, I guess I don't mind, maybe someone else has a preference.

**Comment by kaiyou** [comment]
Two remaining open questions : 
- should we move the Dockerfile to the ``contrib/docker`` folder as well?
- should we rremove the ``SYNAPSE_`` prefix on environment variables?

**Comment by jcgruenhage** [comment]
If you move the dockerfile, how will you pull in the source code?
I'd keep the synapse prefix (changed my mind :D) because if the environment configuration generation is moved into the synapse config generation, people wouldn't have to change their env vars in docker-compose, if this was integrated here.

**Comment by kaiyou** [comment]
Moving the Dockerfile does not mean moving the build context. Docker hub automated build does not support this but a custom build or a DockerCloud build can simply use the ``-f`` flag from ``docker build`` to specify the Dockerfile location.

**Comment by jcgruenhage** [comment]
Ahh right, I forgot about -f. I still would rather keep it in the root, as it is easier to build that way, and as you mentioned, there are environments that don't support `-f`.

**Comment by kaiyou** [comment]
The image is built at ``docker.io/tedomum/synapse`` and is running in production on one of my homeservers. I suggest we wait at least until monday to get some feedback until we think seriously about merging.

**Comment by ara4n** [comment]
this is super-exciting - thanks @kaiyou (and @jcgruenhage) for attacking it. i doubt there is going to be much in the way of review from the core team as I don't think anyone there currently is using docker, but @richvdh/@erikjohnston/me may be able to provide some naive feedback. if you can get more feedback from docker gurus it'd be great to get some more datapoints before we merge :)

**Comment by Gwend4l** [comment]
For information: I'm the docker image provided by @kaiyou (https://hub.docker.com/r/tedomum/synapse/) since Sunday evening, and everything is going well. The configuration through environment variables works great, at least in my case where I'm still using sqlite. Thanks!

**Comment by ara4n** [comment]
@kaiyou should we merge?

**Comment by t3chguy** [comment]
kaiyou says:  I dont have my Github credentials to reply on the thread, but things look good on my server using Postgres and Ive heard from some other happy users, so Id say its safe to merge. Probably some improvements will come later

File: .dockerignore [hunk 27]
```diff
@@ -0,0 +1,5 @@
+Dockerfile
+.travis.yml
+.gitignore
+demo/etc
+tox.ini
```
Comment (dali99) [inline .dockerignore]: This line seems out of place

Comment (kaiyou) [reply .dockerignore]: You mean ``tox.ini``, tests are not run in the Docker image. I guess we could include the tox configuration of course, but I did not find it useful.

Comment (dali99) [reply .dockerignore]: Alright then

**Comment by kaiyou** [comment]
@ara4n, any news about merging so I can start working on additional features for the image?

**Comment by Gwend4l** [comment]
Any update on this front?

**Comment by neilisfragile** [comment]
Hi @Gwend4l @michaelkaye has joined the core team in an SRE capacity - he needs to take a look before merge - but it is on his immediate todo list

**Comment by Gwend4l** [comment]
sounds great! then good luck @michaelkaye :)

**Comment by michaelkaye** [comment]
Hi there @kaiyou  - apologies for it taking so long to get back to you.

Currently we've been doing the packaging of synapse out of the main tree - for instance the Debian build lives over https://github.com/matrix-org/package-synapse-debian in it's own packaging.

I know that makes it a bit more of a pain to package, but at the moment we want to keep things separated. 

I've gone and made a new repository in the matrix-org org to host this : https://github.com/matrix-org/matrix-synapse-docker - if we can move the code over to be a PR against that then I think we're happy to merge it in.

Once we've got the image building, I'd like to get some automation on builds into the docker-hub repos, and we can make doing a commit to bump the version part of our general synapse release process.

Cheers,
Michael

**Comment by michaelkaye** [comment]
It is probably worth also mentioning that I'd like to find a better way for these to be configured than the current (slightly fragile) way of writing a new config file on the fly. It's something that needs a bit of careful thought and probably will involve splitting out the way the configuration parsing is happening, as not everyone is deploying synapse now by installing it then running it immediately for the configuration file. Still, something for the future, that.

File: contrib/docker/README.md [hunk 28]
```diff
@@ -0,0 +1,148 @@
...
+If ``SYNAPSE_CONFIG_PATH`` is set, you should generate a configuration file
+then customize it manually. No other environment variable is required.
+
+Otherwise, a dynamic configuration file will be used. The following environment
+variables are available for configuration:
+
+* ``SYNAPSE_SERVER_NAME`` (mandatory), the current server public hostname.
+* ``SYNAPSE_REPORT_STATS``, (mandatory, ``yes`` or ``no``), enable anonymous
+  statistics reporting back to the Matrix project which helps us to get funding.
+* ``SYNAPSE_MACAROON_SECRET_KEY`` (mandatory) secret for signing access tokens
```
Comment (turt2live) [inline contrib/docker/README.md]: iirc this can't change after initially being set otherwise synapse can't understand previous access tokens.

**Comment by kaiyou** [comment]
I must admin I have a strong opinion about keeping the Dockerfile inside the repository, so it evolves with it and one can build any commit with a simple checkout.

Regarding the configuration being generated on the fly, I find it nice to have both options, which the current Dockerfile does provide, so that inexperienced users can still run an instance with sane defaults for a small-sized server.

File: contrib/docker/README.md (hunk 28 above)
Comment (kaiyou) [reply contrib/docker/README.md]: Yes, I did hesitate with a default behavior generating a random value and storing it to the data dir.

**Comment by jcgruenhage** [comment]
For building the specific commit, you can make it into a build arg in the docker file and pass it when building the container. That even has the advantage that you can build an older version where the Dockerfile wasn't merged.

**Comment by kaiyou** [comment]
Except the Dockerfile depends on the content of the repo itself. Now it is running on Python2, maybe tomorrow Synapse will run on Python3, maybe some dependencies will change and require library headers, specific library versions, etc. My experience with Dockerfiles is: keeping them outside the main repository is a burden.

File: contrib/docker/README.md (hunk 28 above)
Comment (turt2live) [reply contrib/docker/README.md]: Some added documentation to say "don't ever change this" would be nice :)

File: contrib/docker/README.md [hunk 29]
```diff
@@ -0,0 +1,148 @@
...
+* ``SYNAPSE_TURN_SECRET``, set this to the TURN shared secret if required.
+
+Shared secrets, that will be initialized to random values if not set:
+
+* ``SYNAPSE_REGISTRATION_SHARED_SECRET``, secret for registrering users if
+  registration is disable.
+
+Database specific values (will use SQLite if not set):
+
+* `POSTGRES_DATABASE` - The database name for the synapse postgres database. [default: `matrix`]
```
Comment (Ezwen) [inline contrib/docker/README.md]: Shouldn't that be POSTGRES_DB, as in https://github.com/kaiyou/synapse/blob/feat-dockerfile/contrib/docker/conf/homeserver.yaml#L55 ?

Comment (jcgruenhage) [reply contrib/docker/README.md]: If so, the default also seems to be "synapse" too.

**Comment by kaiyou** [comment]
The latest build is affected by #3135, I will fix tonight.

**Comment by almereyda** [comment]
@kaiyou If you're at it, please also consider to include the recent security release https://github.com/matrix-org/synapse/releases/tag/v0.28.1

I'm going to update matrix.allmende.io from 0.26 to your branch in the next days and can provide feedback them.

**Comment by michaelkaye** [comment]
Hey, so having had a bit of a think about what would be good as a whole for packaging, and a chat with a few people, and also seeing the support for in-tree packaging in the comments here.

The logic about "the debian packaging and the docker packaging should be supported to the same level" and be treated similarly still applies - but maybe that actually means bringing both into this repository (or in the case of debian, into a branch in the repository).

What I suppose we were worried about is a free for all for all the varied packaging solutions that are around - we can't afford to support all the options that are out there, so limiting ourselves to tarballs, docker images and debian packages seems sensible - things we want to make sure are pushed as part of a synapse release.

tl;dr:

@kaiyou  - thank you very much for your work and for waiting - Can you have a look at https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst and make an appropriate entry in the PR description, and we'll be able to merge once that's done.

**Comment by michaelkaye** [comment]
I should probably also ask @jcgruenhage to do something about the status of his review - i believe the comments there have been addressed?

**Comment by michaelkaye** [comment]
Probably also worth thinking about what we'll be doing to push these packages. I was thinking to make something visible on docker.io/matrixdotorg/synapse we setup the automated builds there to trigger only when there's a new tag pushed - eg "v*".

If we run into release problems with this, we can just make a little bugfix branch and push the matrixdotorg image out by hand, so they'll be available there.

**Comment by kaiyou** [comment]
Docker cloud will basically build every branch and tag automatically, or you can provide a filtering regex.

**Comment by jcgruenhage** [comment]
@michaelkaye must have forgottten to set it to approved back when I did the review and @kaiyou changed what I wanted
